chain-patrol-ts/src/fetcher.ts

```diff
@@ -0,0 +1,92 @@
+import { LRUCache } from "lru-cache";
+import { createAssetListApiOptions, createAssetDetailsApiOptions } from "./utils";
+import { ApiOptions, AssetList, Asset, AssetDetails } from "./types";
+import { ASSET_TYPES } from "./constants";
+
+export default class AssetFetcher {
+  assetListUrl: string;
+  assetDetailsUrl: string;
+  assetStatus: string;
+  private apiKey: string;
+  private readonly MAX_TRIES: number;
+  private assetListCache: LRUCache<string, Asset[]>;
+  private assetDetailsCache: LRUCache<string, AssetDetails>;
+
+  constructor(apiKey: string, assetListUrl: string, assetDetailsUrl: string, assetStatus: string) {
+    this.MAX_TRIES = 3;
+    this.apiKey = apiKey;
+    this.assetListUrl = assetListUrl;
+    this.assetDetailsUrl = assetDetailsUrl;
+    this.assetStatus = assetStatus;
+    this.assetListCache = new LRUCache<string, Asset[]>({
+      max: 1000,
+    });
+    this.assetDetailsCache = new LRUCache<string, AssetDetails>({
+      max: 1000,
+    });
+  }
+
+  getAssetlist = async (endDate: string, startDate: string): Promise<Asset[] | undefined> => {
+    const key = `${endDate}-${startDate}`;
+    if (this.assetListCache.has(key)) {
+      return this.assetListCache.get(key);
+    }
+
+    let tries = 0;
+    while (tries < this.MAX_TRIES) {
+      try {
+        // Since we can only query the API
+        // with one `type` at a time
+        const apiOptions: ApiOptions[] = [];
+        ASSET_TYPES.map((type: string) => {
+          apiOptions.push(createAssetListApiOptions(this.apiKey, type, this.assetStatus, endDate, startDate));
+        });
+
+        let fetchedAssets: Asset[] = [];
+
+        await Promise.all(
+          apiOptions.map(async (options: ApiOptions) => {
+            const fetchedAssetList = (await (await fetch(this.assetListUrl, options)).json()) as AssetList;
+            fetchedAssets.push(...fetchedAssetList.assets);
+          })
+        );
+        this.assetListCache.set(key, fetchedAssets);
+        return fetchedAssets;
+      } catch (e) {
+        tries++;
+        console.log(`Error in fetching Asset List (attempt ${tries}): `, e);
+        if (tries === this.MAX_TRIES) {
+          throw e;
+        }
+      }
+      // Wait for 1 second before retrying
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    }
+    return undefined;
+  };
+
+  getAssetDetails = async (assetContent: string): Promise<AssetDetails | undefined> => {
+    if (this.assetDetailsCache.has(assetContent)) {
+      return this.assetDetailsCache.get(assetContent);
+    }
+
+    let tries = 0;
+    while (tries < this.MAX_TRIES) {
+      try {
+        const options = createAssetDetailsApiOptions(this.apiKey, assetContent);
+        const assetDetails = (await (await fetch(this.assetDetailsUrl, options)).json()) as AssetDetails;
```

Comment:
since both of these `getAssetlist` and `getAssetDetails` functions look similar repetitive, I think it might be good to create a separate function for fetching and retrying.

here's my first quick draft idea:

```
  async fetchWithRetries(apiUrl: string, apiOptions: ApiOptions, maxRetries: number): Promise<any> {
    let tries = 0;
    while (tries < maxRetries) {
      try {
        return await fetch(apiUrl, apiOptions).then((response) => response.json());
      } catch (e) {
        tries++;
        console.log(`Error in fetching data (attempt ${tries}): `, e);
        if (tries === maxRetries) {
          throw e;
        }
      }
      // Wait for 1 second before retrying
      await new Promise((resolve) => setTimeout(resolve, 1000));
    }
    return undefined;
  }
  
  
  getAssetlist = async (endDate: string, startDate: string): Promise<Asset[] | undefined> => {
    const key = `${endDate}-${startDate}`;
    if (this.assetListCache.has(key)) {
      return this.assetListCache.get(key);
    }

    const apiOptionsList: ApiOptions[] = ASSET_TYPES.map((type) =>
      createAssetListApiOptions(this.apiKey, type, this.assetStatus, endDate, startDate)
    );

    const fetchedAssets = await Promise.all(
      apiOptionsList.map((options) => this.fetchWithRetries(this.assetListUrl, options, this.MAX_TRIES))
    );

    this.assetListCache.set(key, fetchedAssets);
    return fetchedAssets;
  };
  
  
  getAssetDetails = async (assetContent: string): Promise<AssetDetails | undefined> => {
    if (this.assetDetailsCache.has(assetContent)) {
      return this.assetDetailsCache.get(assetContent);
    }

    const options = createAssetDetailsApiOptions(this.apiKey, assetContent);
    const assetDetails = await this.fetchWithRetries(this.assetDetailsUrl, options, this.MAX_TRIES);

    this.assetDetailsCache.set(assetContent, assetDetails);
    return assetDetails;
  };
```

Reply:
Great suggestion, Tayfun 🔥 

I'll work on implementing this